packages/flutter/lib/src/material/bottom_navigation_bar.dart

```diff
@@ -0,0 +1,551 @@
+        reverseCurve: Curves.fastOutSlowIn.flipped
+      );
+    });
+
+    _currentIndex = 0;
+    _controllers[_currentIndex].value = 1.0;
+    if (config.onTap != null)
+      config.onTap(_animations[_currentIndex], _currentIndex);
+
+    _backgroundColor = config.labels[_currentIndex].backgroundColor;
+  }
+
+  @override
+  void dispose() {
+    for (AnimationController controller in _controllers)
+      controller.dispose();
+    for (_Circle circle in _circles)
+      circle.dispose();
+    super.dispose();
+  }
+
+  void _rebuild() {
+    setState(() {
+      // Animation tick.
+    });
+  }
+
+  double get _maxWidth {
+    assert(config.type != null);
+    switch (config.type) {
+      case BottomNavigationBarType.fixed:
+        return config.labels.length * _kActiveMaxWidth;
+      case BottomNavigationBarType.shifting:
+        return _kActiveMaxWidth + (config.labels.length - 1) * _kInactiveMaxWidth;
+    }
+    return null;
+  }
+
+  bool _isAnimating(Animation<double> animation) {
+    return animation.status == AnimationStatus.forward ||
+           animation.status == AnimationStatus.reverse;
+  }
+
+  // Because of non-linear nature of the animations, the animations that are
+  // currently animating might not add up to the flex weight we are expecting.
+  // (1.5 + N - 1, since the max flex that the animating ones can have is 1.5)
+  // This causes instability in the animation when multiple labels are tapped.
+  // To solves this, we always store a weight that normalizes animating
+  // animations such that their resulting flex values will add up to the desired
+  // value.
+  void _computeWeight() {
+    final Iterable<Animation<double>> animating = _animations.where(
+      (Animation<double> animation) => _isAnimating(animation)
+    );
+
+    if (animating.isNotEmpty) {
+      final double sum = animating.fold(0.0, (double sum, Animation<double> animation) {
+        return sum + _ratioTween.evaluate(animation);
+      });
+      _weight = (animating.length + 0.5) / sum;
+    } else {
+      _weight = 1.0;
+    }
+  }
+
+  double _ratio(Animation<double> animation) {
+    if (_isAnimating(animation)) {
+      assert(_weight != null);
+      return _ratioTween.evaluate(animation) * _weight;
+    } else {
+      return _ratioTween.evaluate(animation);
+    }
+  }
+
+  double _xOffset(int index) {
+    double weightSum(Iterable<Animation<double>> animations) {
+      return animations.map(
+        (Animation<double> animation) => _ratio(animation)
+      ).fold(0.0, (double sum, double value) => sum + value);
+    }
+
+    final double allWeights = weightSum(_animations);
+    final double leftWeights = weightSum(_animations.sublist(0, index));
+
+    return (leftWeights + _ratio(_animations[index]) / 2.0) / allWeights;
+  }
+
+  FractionalOffset cirleOffset(int index) {
+    final Tween<double> yOffsetTween = new Tween<double>(
+      begin: (18.0 + 24.0 / 2.0) / kBottomNavigationBarHeight, // 18dp + icon center
```

Comment:
You're assuming that the icon's size is the default (24).  Should DestinationLabel assert if it's given icons with a different size? Maybe you should use the icon's actual size here?